cmd/nitro/init.go

```diff
@@ -0,0 +1,383 @@
+			if storedGenBlock.Header().Root == (common.Hash{}) {
+				panic(fmt.Errorf("Attempting to init genesis block %x, but this block is in database with no state root", genesisBlockNr))
+			}
+			log.Warn("Re-creating genesis though it seems to exist in database", "blockNr", genesisBlockNr)
+		}
+		log.Info("Initializing", "ancients", ancients, "genesisBlockNr", genesisBlockNr)
+		l2BlockChain, err = arbnode.WriteOrTestBlockChain(chainDb, cacheConfig, initDataReader, chainConfig, &config.Node, config.Init.AccountsPerSync)
+		if err != nil {
+			panic(err)
+		}
+	}
+	txIndexWg.Wait()
+	err = chainDb.Sync()
+	if err != nil {
+		panic(err)
+	}
+
+	err = validateBlockChain(l2BlockChain, chainConfig.ChainID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return chainDb, l2BlockChain, nil
+}
+
+func testTxIndexUpdated(chainDb ethdb.Database, lastBlock uint64) bool {
+	var transactions types.Transactions
+	blockHash := rawdb.ReadCanonicalHash(chainDb, lastBlock)
+	reReadNumber := rawdb.ReadHeaderNumber(chainDb, blockHash)
+	if reReadNumber == nil {
+		return false
+	}
+	for ; ; lastBlock-- {
+		blockHash := rawdb.ReadCanonicalHash(chainDb, lastBlock)
+		block := rawdb.ReadBlock(chainDb, blockHash, lastBlock)
+		transactions = block.Transactions()
+		if len(transactions) == 0 {
+			if lastBlock == 0 {
+				return true
+			}
+			continue
+		}
+		entry := rawdb.ReadTxLookupEntry(chainDb, transactions[len(transactions)-1].Hash())
+		return (entry != nil)
+	}
+}
+
+func testUpdateTxIndex(chainDb ethdb.Database, chainConfig *params.ChainConfig, txIndexWg *sync.WaitGroup) {
+	lastBlock := chainConfig.ArbitrumChainParams.GenesisBlockNum
+	if lastBlock == 0 {
+		// no Tx, no need to update index
+		return
+	}
+
+	lastBlock -= 1
+	if testTxIndexUpdated(chainDb, lastBlock) {
+		return
+	}
+
+	txIndexWg.Add(1)
+	log.Info("writing Tx lookup entries")
+
+	go func() {
+		batch := chainDb.NewBatch()
+		for blockNum := uint64(0); blockNum <= lastBlock; blockNum++ {
+			blockHash := rawdb.ReadCanonicalHash(chainDb, blockNum)
+			block := rawdb.ReadBlock(chainDb, blockHash, blockNum)
+			txs := block.Transactions()
+			txHashes := make([]common.Hash, 0, len(txs))
+			txHashMap := make(map[common.Hash]int, len(txs))
+			for _, tx := range txs {
+				txHash := tx.Hash()
+				txHashes = append(txHashes, txHash)
+				txHashMap[txHash]++
+			}
+			for txHash := range txHashMap {
+				if entry := rawdb.ReadTxLookupEntry(chainDb, txHash); entry != nil {
+					txHashMap[txHash]++
+				}
+			}
+			var receipts types.Receipts = nil
+			for i, txHash := range txHashes {
+				if txHashMap[txHash] > 1 {
+					if receipts == nil {
+						receipts = rawdb.ReadReceipts(chainDb, blockHash, blockNum, chainConfig)
+					}
+					if receipts[i].Status == 0 && receipts[i].GasUsed == 0 {
+						log.Info("Not indexing failed duplicate transaction", "block", blockNum, "txHash", txHash, "index", i)
+						txHashes[i] = common.Hash{}
+						txHashMap[txHash]--
```

Comment:
I think we should instead remove this from the txHashes list (or rather create a new txHashes list without the duplicates), so that we don't create a tx lookup entry for the zero hash